src/music/MusicManager.ts

```diff
@@ -0,0 +1,185 @@
+        )
+    };
+
+    export const masterGain: GainNode = context.createGain();
+    export const drumkit: Drumkit = new Drumkit();
+
+    interface MusicState {
+        beatsPerMinute: number;
+        rhythm: Rhythm;
+        root: number;
+        scale: Scale;
+    }
+
+    const state: MusicState = {
+        beatsPerMinute: 220,
+        rhythm: new Rhythm(),
+        root: 60, // middle C
+        scale: 2741 // major
+    };
+
+    enum ChordFunction {
+        TONIC = "tonic",
+        SUBDOMINANT = "subdominant",
+        DOMINANT = "dominant",
+        AMBIGUOUS = "ambiguous"
+    }
+
+    const backHalves: NonEmptyArray<ChordFunction[]> = [
+        [ChordFunction.TONIC, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.AMBIGUOUS, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT, ChordFunction.AMBIGUOUS]
+    ];
+
+    const FunctionDegrees: Record<ChordFunction, NonEmptyArray<number>> = {
+        "tonic": [0, 5], // I and VI
+        "subdominant": [1, 3], // II and IV
+        "dominant": [4, 6], // V and VII
+        "ambiguous": [2] // III
+    };
+
+    function generateChordProgression(scale: Scale, root: number, extensions: number = 4): number[][] {
+        const pitchClass: number[] = Scales.getPitchClass(scale);
+        const backHalf: ChordFunction[] = Random.fromArray(backHalves);
+        // always start from the tonic and add a valid back half
+        // this way it'll always fit the tonic -> subdominant -> dominant pattern
+        const chordDegrees: number[] = [0].concat(backHalf.map(f => Random.fromArray(FunctionDegrees[f])));
+        return chordDegrees.map(degree => {
+            const chord = [];
+            for (let i = 0; i < extensions; i++) {
+                // step up scale in thirds starting from chord's root
+                chord[i] = Scales.indexIntoPitchClass(pitchClass, degree + 2 * i) + root;
+            }
+            return chord;
+        });
+    }
+
+    function generateDrumLoop(): Drums[] {
+        const drumLoop: Drums[] = [];
+        const subdivision: number[] = state.rhythm.subdivision;
+        let onKick: boolean = true;
+        for (let i = 0; i < subdivision.length; i++) {
+            let acc: number = subdivision[i] - 1;
+            if (onKick) {
+                drumLoop.push(Drums.KICK);
+            } else {
+                drumLoop.push(Drums.SNARE);
+            }
+            onKick = !onKick;
+            while (acc > 0) {
+                if (acc === 1 && i === subdivision.length - 1) {
+                    drumLoop.push(Drums.OPEN_HI_HAT);
+                } else {
+                    drumLoop.push(Drums.CLOSED_HI_HAT);
+                }
+                acc--;
+            }
+        }
```

Comment:
This could use a lot of style cleanup. Also, explain what the process is here, in docs above the function. You seem to make several arbitrary but undocumented decisions about where to put what.

Reply:
Code's been restructured, and there's some explanatory comments inside the code. Please let me know if it needs more documentation.